[Heartbeat] failure in unit test heartbeat/scheduler TestSchedTaskLimits/runs_50_interleaved_without_limit

## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2Fmain/detail/main/1522//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2Fmain/detail/main/1522//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2Fmain/detail/main/1522//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2Fmain/detail/main/1522//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://beats_null.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2023-05-15T09:57:14.185Z&rangeTo=2023-05-15T10:17:14.185Z&transactionName=BUILD+Beats%2Fbeats%2Fmain&transactionType=job&latencyAggregationType=avg&traceId=8dc814c5a9603e27cf384d1d782a8fcf&transactionId=ec8634689010a7ad)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2023-05-15T10:07:14.185+0000


* Duration: 128 min 9 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 51218  |
| Skipped      | 4798 |
| Total        | 56016   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->














### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Beats%2Fbeats%2Fmain/detail/main/1522//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `heartbeat-windows-8-windows-8 - mage build unitTest`
<ul>
<li>Took 2 min 11 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/main/runs/1522/steps/30474/log/?start=0">here</a></li>
<li>Description: <code>mage build unitTest</code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 12 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Beats/pipelines/beats/pipelines/main/runs/1522/steps/26194/log/?start=0">here</a></li>
<li>Description: <code>[2023-05-15T11:47:57.866Z] [Google Cloud Storage Plugin] Found 1 files to download from pattern: gs:</code></l1>
</ul>
  
  </p>
  </details>

Failure in heartbeat unit tests

```
[2023-05-15T11:53:43.660Z] FAIL
[2023-05-15T11:53:43.661Z] FAIL	github.com/elastic/beats/v7/heartbeat/scheduler	0.313s
...
[2023-05-15T11:53:50.055Z] === Failed
[2023-05-15T11:53:50.055Z] === FAIL: heartbeat/scheduler TestSchedTaskLimits/runs_50_interleaved_without_limit (0.00s)
[2023-05-15T11:53:50.055Z]     testing.go:1343: test executed panic(nil) or runtime.Goexit: subtest may have called FailNow on a parent test
[2023-05-15T11:53:50.055Z]     --- FAIL: TestSchedTaskLimits/runs_50_interleaved_without_limit (0.00s)
[2023-05-15T11:53:50.055Z] 
[2023-05-15T11:53:50.055Z] === FAIL: heartbeat/scheduler TestSchedTaskLimits (0.00s)
[2023-05-15T11:53:50.056Z]     scheduler_test.go:234: 
[2023-05-15T11:53:50.056Z]         	Error Trace:	C:\Users\jenkins\workspace\main-1522-7be53732-1af6-428f-952e-92bc0cae370d\src\github.com\elastic\beats\heartbeat\scheduler\scheduler_test.go:234
[2023-05-15T11:53:50.056Z]         	            				C:\Users\jenkins\workspace\main-1522-7be53732-1af6-428f-952e-92bc0cae370d\src\github.com\elastic\beats\heartbeat\scheduler\scheduler_test.go:271
[2023-05-15T11:53:50.056Z]         	Error:      	"37" is not greater than or equal to "50"
[2023-05-15T11:53:50.056Z]         	Test:       	TestSchedTaskLimits
```

--- a/heartbeat/scheduler/scheduler_test.go
+++ b/heartbeat/scheduler/scheduler_test.go
@@ -231,7 +231,7 @@ func TestSchedTaskLimits(t *testing.T) {
 			numJobs: 50,
 			limit:   math.MaxInt64,
 			expect: func(events []int) {
-				require.GreaterOrEqual(t, len(events), 50)
+				require.GreaterOrEqual(t, len(events), 250)
 			},
 		},
 		{
@@ -239,7 +239,7 @@ func TestSchedTaskLimits(t *testing.T) {
 			numJobs: 100,
 			limit:   0,
 			expect: func(events []int) {
-				require.GreaterOrEqual(t, len(events), 100)
+				require.GreaterOrEqual(t, len(events), 500)
 			},
 		},
 	}
@@ -254,11 +254,15 @@ func TestSchedTaskLimits(t *testing.T) {
 			}
 			s := Create(math.MaxInt64, monitoring.NewRegistry(), tarawaTime(), jobConfigByType, false)
 			var taskArr []int
+			mtx := sync.Mutex{}
 			wg := sync.WaitGroup{}
 			wg.Add(tt.numJobs)
 			for i := 0; i < tt.numJobs; i++ {
 				num := i
 				tf := makeTasks(4, func() {
+					mtx.Lock()
+					defer mtx.Unlock() // taskArr is shared across goroutines
+
 					taskArr = append(taskArr, num)
 				})
 				go func(tff TaskFunc) {

